src/main/scala/org/dbpedia/spotlight/wikistats/ComputeStats.scala

```diff
@@ -17,50 +17,134 @@
-    val dfSurfaceForms = dfWikiRDD.select("links.id")//.where("")
-                        .rdd
-                        .map(artRow => artRow.getList[String](0))
-                        .flatMap(articleIds => articleIds.map(id=>id))
-                        .mapPartitions{ wikiIds =>
-                                        val dbpediaEncode = new DBpediaUriEncode(language)
-                                        wikiIds.map(wikiId => (dbpediaEncode.uriEncode(wikiId),1))}
-                        .reduceByKey(_ + _)
+    val allSfs = wikipediaParser.getSfs().collect().toList
 
+    //Below Logic is to get Tokens from the list of Surface forms
+    val tokens = wikipediaParser.getTokens()
 
+    //Creating MemoryTokenTypeStore from the list of Tokens
+    val tokenTypeStore = wikipediaParser.createTokenTypeStore(tokens)
+
+    //Broadcast TokenTypeStore for creating tokenizer inside MapPartitions
+    val tokenTypeStoreBc = sc.broadcast(tokenTypeStore)
+
+    val langTokenizer = new LanguageTokenizer(lang)
+    val lit = langTokenizer.litInstance(tokenTypeStore)
+
+    //Creating dictionary broadcast
+    val fsaDict = FSASpotter.buildDictionaryFromIterable(allSfs,lit)
+    val fsaDictBc = sc.broadcast(fsaDict)
 
-    dfSurfaceForms.foreach(println)
-  }
 
-  def sfCounts(dfWikiRDD:DataFrame): Unit={
+    //Get wid and articleText for FSA spotter
+    val textIdRDD = wikipediaParser.getArticleText()
 
+    //textIdRDD.foreach(println)
+    //Implementing the FSA Spotter logic
+
+    //Declaring value for avoiding the whole class to be serialized
+    val language = lang
+
+    import sqlContext.implicits._
+    //Logic to get the Surface Forms from FSA Spotter
+    val totalSfsRDD = textIdRDD.mapPartitions(textIds => {
+              textIds.map(textId => {
+                      val langTokenizer = new LanguageTokenizer(language)
+                      val allOccFSASpotter = new AllOccurrencesFSASpotter(fsaDictBc.value,
+                                                                          langTokenizer.litInstance(tokenTypeStoreBc.value))
+                      allOccFSASpotter
+                      .extract(textId._2)
+                      .map(sfOffset => (textId._1,sfOffset._1))
+
+              })
+              .flatMap(idSf => idSf)
+    })
+
+    val totalSfDf = totalSfsRDD.toDF("wid", "sf2")
+    val uriSfDf = wikipediaParser.getSfURI().toDF("wid", "sf1", "uri")
+
+    counts(uriSfDf,totalSfDf)
   }
 
   /*
-
-  TODO All the other Counts methods here.
+  Method to Compute various counts on the WikiDump
    */
+  def counts(uriSfDf:DataFrame, totalSfDf:DataFrame) = {
 
+    //Joining two Datasets
+    val joinedDf = uriSfDf.join(totalSfDf,(uriSfDf("wid") === totalSfDf("wid"))
+                                && (uriSfDf("sf1") === totalSfDf("sf2")),"left_outer")
+
+    val language = lang
+    //Get the URI Counts
+
+    val uriCounts = joinedDf
+                    .groupBy("uri")
```

Comment:
I will check this. Groupby is a datarame operation..like a sql groupby. Not sure whether df has reduceby.